### PR TITLE
Update Metadata.java

### LINKED ISSUE #699: MapDB: Not prepared to work with Metadata?
I'm working with item metadata in https://github.com/openhab/openhab2-addons/pull/5243 (Hue Emulation Service) and I'm using https://github.com/maggu2810/openhab-demo as the runtime for development. That runtime is configured to use mapdb instead of jsondb which exposes the following deserialize problem:

```
o.e.s.s.mapdb.internal.MapDbStorage:171 - Couldn't deserialize value 'org.eclipse.smarthome.core.items.Metadata@@@{"key":{"segments":["HUEEMU","DemoDimmer"]},"value":"5","configuration":{}}'. Root cause is: Unable to invoke no-args constructor for class org.eclipse.smarthome.core.items.Metadata. Registering an InstanceCreator with Gson for this type may fix this problem.
```

Afaik mapdb is still a supported db. Can someone confirm the above issue?

Quick demo code:

```java
@Component(immediate = true)
@NonNullByDefault
public class MyClass {
    @Reference
    protected @NonNullByDefault({}) MetadataRegistry metadataRegistry;

    @Activate
    public void activate(Map<String, Object> properties) {
         MetadataKey key = new MetadataKey("some_namespace", "existing_item_name");
         Metadata metadata = metadataRegistry.get(key);

         if (metadata != null) {
              metadataRegistry.add(new Metadata(key, "a_value", null));
         }
    }
}
```

Update: I have changed to jsondb and I'm experiencing the same problem. My next guess is that due to the update to a newer (and stricter!) gson library, gson can no longer (de)serialize classes with a no-args constructor. And MetaData does not have that!

### PR DESCRIPTION
**Pull request by davidgraeff**
Fixes #699 

Signed-off-by: David Gräff <david.graeff@web.de>

### CODE REVIEW

File: bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java [hunk 1]
```diff
@@ -35,6 +35,9 @@
     private final String value;
     private final Map<String, Object> configuration;
 
+    Metadata() { // For (de)serialization
```
Comment (maggu2810) [inline bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java]: Can't we use the same comment for all such constructors?

File: bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java [hunk 2]
```diff
@@ -35,6 +35,9 @@
...
+    }
```
Comment (maggu2810) [inline bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java]: You need to assign a "dummy" value to the final fields to pass the null checks (all members are non null).
You will see an example in one of the referenced PR.

Comment (davidgraeff) [reply bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java]: True. Regarding this other branch, I'm using the github editor to get this done as another core checkout and build would poison my local maven cache.

Comment (davidgraeff) [reply bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java]: fixed. But you are rather verbose here in terms of LOC for this actually very obvious fact. I think I liked my one liner more.

Comment (maggu2810) [reply bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java]: It has not been chosen by me (IIRC) but present in some ESH file for years.
My requirement is that it is the same for all files.

So, if we would like to search all files and replace it, I am still happy. :wink:

Comment (maggu2810) [reply bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java]: > as another core checkout and build would poison my local maven cache.

Just if you are interested in: `mvn -Dmaven.repo.local=/some/other/directory ...`

Comment (davidgraeff) [reply bundles/org.openhab.core/src/main/java/org/eclipse/smarthome/core/items/Metadata.java]: I'm a gradle user and so to say "forced" to use maven for this project. I turns out it is not 10 times slower as I originally thought (that was tycho) and if you turn it silent, it also doesn't cause so much noise anymore. But I'm still not familiar with all those command line switches. Thanks for the hint.

**Comment by maggu2810** [comment]
I tested the build locally.

**Pull request merged**